### PR TITLE
DURACOM-86 Add html support to opensearch, fix misconfiguration

### LINKED ISSUE #2869: DS-4249: Add support for HTML response in OpenSearch
This is a duplicate of [DS-4249](https://jira.lyrasis.org/browse/DS-4249)

Estimate from release plan: 2 hours
Expressing interest: none

This is a followup of DS-3943, more specifically the merged PR
https://github.com/DSpace/DSpace/pull/2128
doesn't include support for

Handle HTML responses in OpenSearchController or correct link options in dspace.cfg and/or https://github.com/DSpace/DSpace/blob/master/dspace-api/src/main/java/org/dspace/app/util/OpenSearchServiceImpl.java to redirect requests to basic search module of the UI

### PR DESCRIPTION
**Pull request by abollini**
## References
* Fixes #2869

## Description
Other than configuring by default the html format for opensearch, redirecting to the standard search module as in previous dspace versions, this PR address a small configuration bug that leads to a wrong opensearch service document.

## Instructions for Reviewers
Enable the opensearch support and look to the opensearch service document before and after applying this PR.
You should note the following changes:
* the html format should be listed in the service document and the URL should point to the public search
* the other formats (atom, rss) should include the proper link to the opensearch controller (before it contains /null)
* run an empty opensearch search will not longer throws a NPE ${dspace.url}/server/opensearch/search

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [n/a] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [n/a] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [n/a] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [n/a] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace/config/dspace.cfg [hunk 1]
```diff
@@ -1316,7 +1316,9 @@ websvc.opensearch.enable = true
 # url used in service document
 websvc.opensearch.svccontext = opensearch
 # context for html request URLs - change only for non-standard servlet mapping
-websvc.opensearch.uicontext = simple-search
+websvc.opensearch.uicontext = search
+# context for xml request URLs - change only for non-standard servlet mapping
+websvc.opensearch.svccontext = opensearch/search
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:1321]: This setting appears twice now in `dspace.cfg` with two different values. It's on line 1317 and here.  I suspect lines 1316-1317 should be removed.

File: dspace/config/dspace.cfg [hunk 2]
```diff
@@ -1335,8 +1337,8 @@ websvc.opensearch.samplequery = photosynthesis
 # tags used to describe search service
 websvc.opensearch.tags = IR DSpace
 # result formats offered - use 1 or more comma-separated from: html,atom,rss
-# NB: html is not supported in DSpace7, use normal search module instead
-websvc.opensearch.formats = atom,rss
+# html uses the normal search module
+websvc.opensearch.formats = html,atom,rss
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:1341]: I'm not able to figure out how to get HTML format to work properly.  It looks like we have a disabled HTML test here: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/opensearch/OpenSearchControllerIT.java#L60 

However, if I enable that test, it fails.

File: dspace/config/dspace.cfg [hunk 3]
```diff
@@ -1316,7 +1316,9 @@ websvc.opensearch.enable = true
 # url used in service document
 websvc.opensearch.svccontext = opensearch
 # context for html request URLs - change only for non-standard servlet mapping
-websvc.opensearch.uicontext = simple-search
+websvc.opensearch.uicontext = search
```
Comment (tdonohue) [inline dspace/config/dspace.cfg:1319]: This change appears to imply you expect `websvc.opensearch.uicontext` to be relative to the `dspace.ui.url`.  Unfortunately, though, currently in the codebase it's relative to `dspace.server.url`:  https://github.com/DSpace/DSpace/blob/main/dspace-api/src/main/java/org/dspace/app/util/OpenSearchServiceImpl.java#L98-L101

Maybe there's another bug in the codebase, or I'm misunderstanding this change?

**Review by tdonohue — CHANGES_REQUESTED** [review]
@abollini : I'm not able to get this to work properly.  It's possible it's a misunderstanding on my end, but I don't see how these simple changes bring HTML support to OpenSearch.  As I note in my inline comments below, we have an `@Ignored` Test which should be enabled when HTML support works for OpenSearch, but it still fails even with these changes.

It's also possible I just don't understand how best to test this.  I expected to be able to do something like this `http://localhost:8080/server/opensearch/search?query=test&format=html`.  But, it doesn't work...it just fails with an 400 exception.

Could you verify how to test this & respond to my inline comments below? As noted, I feel like I'm misunderstanding this PR, but it's also possible that it's simply incomplete.

File: dspace/config/dspace.cfg (hunk 1 above)
Comment (abollini) [reply dspace/config/dspace.cfg:1319]: right. Duplicates have been removed

File: dspace/config/dspace.cfg (hunk 2 above)
Comment (abollini) [reply dspace/config/dspace.cfg:1339]: thanks to spot that. Indeed, the test is wrong. The html format is managed by the UI, so the test should be eventually replaced by an e2e test. I have removed the test and added a comment about it

File: dspace/config/dspace.cfg (hunk 3 above)
Comment (abollini) [reply dspace/config/dspace.cfg:1317]: you are right, I fixed the code to look to the right configuration (as it is the **UI**context)

**Comment by abollini** [comment]
The easier way to test this PR is to look to the opensearch service document description produced at http://localhost:8080/server/opensearch/service a text/html Url element must be present pointing to the angular search ui

I have extended the servicedocument test to verify that all the Url elements are created as expected.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @abollini !  I've retested this with the changes and it now looks correct.  I appreciate that you also corrected to test for `/opensearch/service` to verify that the returned XML service document is corrected.  I also manually tested that and found that it now correctly reports that the HTML search option is at `${dspace.ui.url}/search?query=[searchTerms]`.

**Pull request merged**